unit_tests/src/test_atomic_apply_acl_operations.cpp

```diff
@@ -177,6 +178,85 @@ TEST_CASE("atomic_apply_acl_operations")
         REQUIRE(contains_error_information(json_error_string));
     }
+
+    SECTION("User and zone must match")
+    {
+        const auto* const test_zone_name = "atomic_acl_test_zone";
+        REQUIRE_NOTHROW(ua::client::add_zone(conn, test_zone_name));
+        const auto zone_info = ua::client::zone_info(conn, test_zone_name);
+        REQUIRE(zone_info.has_value());
+
+        irods::at_scope_exit remove_test_zone{[&conn, &test_zone_name] {
+            if (ua::client::zone_exists(conn, test_zone_name)) {
+                ua::client::remove_zone(conn, test_zone_name);
+            }
+        }};
+
+        ua::user user_a{"atomic_acl_zone_user", test_zone_name};
+
+        irods::at_scope_exit remove_user{[&] { ua::client::remove_user(conn, user_a); }};
+
+        if (!ua::client::exists(conn, user_a)) {
+            ua::client::add_user(conn, user_a, ua::user_type::rodsuser, ua::zone_type::remote);
+        }
+
+        REQUIRE(ua::client::exists(conn, user_a));
+
+        irods::at_scope_exit remove_acls{[&] {
+            const auto json_input =
+                json{{"logical_path", user_home},
+                     {"operations",
+                      json::array({{{"entity_name", user_a.name}, {"zone", test_zone_name}, {"acl", "null"}}})}}
+                    .dump();
+
+            char* json_error_string{};
+            irods::at_scope_exit free_memory{[&json_error_string] { std::free(json_error_string); }};
+
+            REQUIRE(rc_atomic_apply_acl_operations(
+                        static_cast<rcComm_t*>(conn), json_input.c_str(), &json_error_string) == 0);
+            REQUIRE(json_error_string == "{}"s);
+
+            const auto perms = fs::client::status(conn, user_home).permissions();
+            auto b = std::begin(perms);
+            auto e = std::end(perms);
+
+            REQUIRE(std::find_if(b, e, [&user_a](const fs::entity_permission& p) {
+                        return p.name == user_a.name && p.prms == fs::perms::own;
+                    }) == e);
+        }};
+
+        const auto failing_json_input =
+            json{{"logical_path", user_home},
+                 {"operations", json::array({{{"entity_name", user_a.name}, {"acl", "own"}}})}}
+                .dump();
+
+        const auto success_json_input =
+            json{
+                {"logical_path", user_home},
+                {"operations", json::array({{{"entity_name", user_a.name}, {"zone", test_zone_name}, {"acl", "own"}}})}}
+                .dump();
+
+        char* json_error_string{};
+        irods::at_scope_exit free_memory{[&json_error_string] { std::free(json_error_string); }};
+
+        // should fail: remote zone with no zone param
+        REQUIRE_FALSE(rc_atomic_apply_acl_operations(
+                          static_cast<rcComm_t*>(conn), failing_json_input.c_str(), &json_error_string) == 0);
+        REQUIRE(json_error_string != "{}"s);
```

Comment:
Any best practices for when the message is extremely verbose?

Reply:
Not really. What does the JSON output look like when the API call fails?

Reply:
The message is an exception with full backtrace, so it's pretty hefty.

Reply:
Please post the JSON output so we have something to comment on.

Reply:
```
{"error_message":"iRODS Exception:\n    file: /irods_source/plugins/api/src/
  atomic_apply_acl_operations.cpp\n    function: void (anonymous namespace)::
  throw_if_invalid_entity_id((anonymous namespace)::id_type)\n    line: 204\n
  code: -130000 (SYS_INVALID_INPUT_PARAM)\n    message:\n        Invalid
  entity\nstack trace:\n--------------\n 0# irods::stacktrace::dump[abi:cxx11]
  () const in /lib/libirods_common.so.4.90.0\n 1# irods::exception::
  assemble_full_display_what() const in /lib/libirods_common.so.4.90.0\n 2#
  irods::exception::what() const in /lib/libirods_common.so.4.90.0\n 3#
  0x000073B59D5916B2 in /usr/lib/irods/plugins/api/
  libirods_atomic_apply_acl_operations_server.so\n 4# 0x000073B59D58F006 in /
  usr/lib/irods/plugins/api/libirods_atomic_apply_acl_operations_server.so\n 5#
  irods::experimental::catalog::execute_transaction(nanodbc::connection&amp;, std::
  function&lt;int (nanodbc::transaction&amp;)>) in /lib/libirods_server.so.4.90.0\n 6#
  0x000073B59D58D6A2 in /usr/lib/irods/plugins/api/
  libirods_atomic_apply_acl_operations_server.so\n 7# irods::api_call_adaptor
  &lt;BytesBuf*, BytesBuf**>::operator()(irods::plugin_context&amp;, RsComm*,
  BytesBuf*, BytesBuf**) in /usr/lib/irods/plugins/api/
  libirods_atomic_apply_acl_operations_server.so\n 8# std::_Function_handler
  &lt;irods::error (irods::plugin_context&amp;, RsComm*, BytesBuf*, BytesBuf**),
  irods::api_call_adaptor&lt;BytesBuf*, BytesBuf**> >::_M_invoke(std::_Any_data
  const&amp;, irods::plugin_context&amp;, RsComm*&amp;&amp;, BytesBuf*&amp;&amp;, BytesBuf**&amp;&amp;) in /
  usr/lib/irods/plugins/api/libirods_atomic_apply_acl_operations_server.so\n 9#
  int irods::api_entry::call_handler&lt;BytesBuf*, BytesBuf**>(RsComm*, BytesBuf*,
  BytesBuf**) in /usr/lib/irods/plugins/api/
  libirods_atomic_apply_acl_operations_server.so\n10# rsApiHandler(RsComm*,
  int, BytesBuf*, BytesBuf*) in /lib/libirods_server.so.4.90.0\n11#
  readAndProcClientMsg(RsComm*, int) in /lib/libirods_server.so.4.90.0\n12#
  agentMain(RsComm*) in /lib/libirods_server.so.4.90.0\n13#
  runIrodsAgentFactory(sockaddr_un) in /lib/libirods_server.so.4.90.0\n14#
  main::$_5::operator()() const at rodsServer.cpp:?\n15# main in /usr/sbin/
  irodsServer\n16# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6\n17#
  _start in /usr/sbin/irodsServer\n\n","operation":{"acl":"own","entity_name":
  "atomic_acl_remote_zone_user"},"operation_index":0}
  ```

Reply:
Yikes. Maybe we just want to assert that there is an "error_message" member and that it contains `SYS_INVALID_INPUT_PARAM` and/or "Invalid entity"? Don't need to match all of those contents...

Reply:
That stacktrace is just noise. We should change the API to use `.client_display_what()` instead of `.what()`.

I agree with @alanking on the what needs to be asserted. The error code is very important. The contents of the message, a nice to have so we can detect when things change.

Reply:
How does my most recent approach look?

Reply:
After some discussion, please do the following:
- Change the JSON output structure to use `client_display_what()` instead of `what()`
- Update unit test to check the JSON output structure using function at the bottom of the unit test source file